pkg/hub/addon/discovery_controller.go

```diff
@@ -172,43 +172,35 @@ func (c *addOnFeatureDiscoveryController) syncCluster(ctx context.Context, clust
 	}
 
 	// build labels for existing addons
-	addOnLabels := map[string]string{}
+	addOnLabels := cluster.Labels
 	addOns, err := c.addOnLister.ManagedClusterAddOns(clusterName).List(labels.Everything())
 	if err != nil {
 		return fmt.Errorf("unable to list addOns of cluster %q: %w", clusterName, err)
 	}
 	for _, addOn := range addOns {
+		key := fmt.Sprintf("%s%s", addOnFeaturePrefix, addOn.Name)
+
 		// addon is deleting
 		if !addOn.DeletionTimestamp.IsZero() {
-			continue
+			addOnLabels[fmt.Sprintf("%s-", key)] = ""
 		}
-		key := fmt.Sprintf("%s%s", addOnFeaturePrefix, addOn.Name)
+
 		addOnLabels[key] = getAddOnLabelValue(addOn)
 	}
 
-	// remove addon lable if its corresponding addon no longer exists
-	for key := range cluster.Labels {
-		if !strings.HasPrefix(key, addOnFeaturePrefix) {
-			continue
-		}
-
-		if _, ok := addOnLabels[key]; !ok {
-			addOnLabels[fmt.Sprintf("%s-", key)] = ""
-		}
+	// build cluster labels patch
+	patchBytes, err := json.Marshal(map[string]interface{}{
+		"metadata": map[string]interface{}{
+			"labels": addOnLabels,
+		},
+	})
+	if err != nil {
+		return fmt.Errorf("failed to create patch for cluster %s: %w", cluster.Name, err)
 	}
 
-	// merge labels
-	modified := false
-	cluster = cluster.DeepCopy()
-	resourcemerge.MergeMap(&modified, &cluster.Labels, addOnLabels)
-
-	// no work if the cluster labels have no change
-	if !modified {
```

Comment:
How about keeping the modified check, after the modified check, we create `patchBytes` and apply the patch, thus we can avoid to path the same thing

Reply:
added label equality check, if they are equal, no need to patch.